File: C/2/8.c
#include <stdio.h>

/**
 * A function that returns the value of the integer x rotated to the right by n
 * positions.
 *
 * @param unsigned x Number to have its bits rotated.
 * @param int      n Amount of bits to be rotated.
 *
 * @return unsigned  Number resultant from operation.
 */
unsigned rightrot(unsigned x, int n)
{
    int n_bits = 0;
    unsigned k = x;

    /**
     * We create a mask to keep rightmost n bits.
     */
    unsigned rightmostMask = ~(~0 << n);
    
    if (x == 0) {
        return 0;
    }

    /**
     * We nee to count how many bits 'x' uses.
     */
    while (k) {
        k /= 2;
        n_bits++;
    }

    /**
     * We apply the mask to keep rightmost n bits of x, then shift them left by
     * a factor of how many bytes were kept after the move (i.e. if x uses 7
     * bits and is shifted right by 3, we shift to the left by 3).
     */
    return ((x & rightmostMask) << (n_bits - (n % n_bits))) | (x >> n);
}

/**
 * A program to test changing bits.
 */
main()
{
    printf("rightrot(255, 1)\t=>\t%u\n", rightrot(255, 1));
    printf("rightrot(0, 4)\t\t=>\t%u\n", rightrot(0, 4));
    printf("rightrot(85, 4)\t\t=>\t%u\n", rightrot(85, 4));
    printf("rightrot(~0, 4)\t\t=>\t%u\n", rightrot(~0, 4));
    printf("rightrot(~0, 31)\t=>\t%u\n", rightrot(~0, 31));
    printf("rightrot(10, 3)\t\t=>\t%u\n", rightrot(10, 3));
}

File: rustlang/2/8.rs
/// A function that returns the value of the integer x rotated to the right by n
/// positions.
///
/// @param u32 x Number to have its bits rotated.
/// @param i32 n Amount of bits to be rotated.
///
/// @return u32  Number resultant from operation.
fn rightrot(x: u32, n: i32) -> u32
{
    let mut n_bits = 0;
    let mut k = x;

    let rightmost_mask = !(!0 << n);
    
    if x == 0 {
        return 0;
    }

    while k > 0 {
        k /= 2;
        n_bits += 1;
    }

    return ((x & rightmost_mask) << (n_bits - (n % n_bits))) | (x >> n);
}

/// A program to test changing bits.
fn main()
{
    print!("rightrot(255, 1)\t=>\t{}\n", rightrot(255, 1));
    print!("rightrot(0, 4)\t\t=>\t{}\n", rightrot(0, 4));
    print!("rightrot(85, 4)\t\t=>\t{}\n", rightrot(85, 4));
    print!("rightrot(!0, 4)\t\t=>\t{}\n", rightrot(!0, 4));
    print!("rightrot(!0, 31)\t=>\t{}\n", rightrot(!0, 31));
    print!("rightrot(10, 3)\t\t=>\t{}\n", rightrot(10, 3));
}
